pykafka/rdkafka/_rd_kafkamodule.c

```diff
@@ -0,0 +1,448 @@
+    NULL,  // TODO docstring
+    Message_fields,
+    4
+};
+
+
+/**
+ * Consumer type
+ */
+
+
+typedef struct {
+    PyObject_HEAD
+    rd_kafka_t *rdk_handle;
+    rd_kafka_conf_t *rdk_conf;
+    rd_kafka_queue_t *rdk_queue_handle;
+    rd_kafka_topic_t *rdk_topic_handle;
+    rd_kafka_topic_conf_t *rdk_topic_conf;
+    PyObject *partition_ids;
+} Consumer;
+
+
+static PyObject *
+Consumer_stop(Consumer *self, PyObject *args) {
+    // Call stop on all partitions, then destroy all handles
+
+    PyObject *retval = Py_None;
+    if (self->rdk_topic_handle != NULL) {
+        Py_ssize_t i, len = PyList_Size(self->partition_ids);
+        for (i = 0; i != len; ++i) {
+            // Error handling here is a bit poor; we cannot bail out directly
+            // if we want to clean up as much as we can.  TODO logging
+            long part_id = PyInt_AsLong(PyList_GetItem(self->partition_ids, i));
+            if (part_id == -1) {
+                retval = NULL;
+                continue;
+            }
+            if (-1 == rd_kafka_consume_stop(self->rdk_topic_handle, part_id)) {
+                set_PyRdKafkaError(rd_kafka_errno2err(errno), NULL);
+                retval = NULL;
+                continue;
+            }
+        }
+        Py_CLEAR(self->partition_ids);
+        rd_kafka_topic_destroy(self->rdk_topic_handle);
+        self->rdk_topic_handle = NULL;
+    }
+    if (self->rdk_queue_handle != NULL) {
+        rd_kafka_queue_destroy(self->rdk_queue_handle);
+        self->rdk_queue_handle = NULL;
+    }
+    if (self->rdk_handle != NULL) {
+        rd_kafka_destroy(self->rdk_handle);
+        self->rdk_handle = NULL;
+    }
+    if (self->rdk_conf) {
+        rd_kafka_conf_destroy(self->rdk_conf);
+        self->rdk_conf = NULL;
+    }
+    if (self->rdk_topic_conf) {
+        rd_kafka_topic_conf_destroy(self->rdk_topic_conf);
+        self->rdk_topic_conf = NULL;
+    }
+    Py_XINCREF(retval);
+    return retval;
+}
+
+
+static void
+Consumer_dealloc(PyObject *self) {
+    PyObject *stop_result = Consumer_stop((Consumer *)self, NULL);
+    if (!stop_result) {
+        // TODO log exception but do not re-raise
+    } else {
+        Py_DECREF(stop_result);
+    }
+    self->ob_type->tp_free(self);
+}
+
+
+static const char Consumer_configure__doc__[] =
+"Set up and populate the rd_kafka_(topic_)conf_t\n"
+"\n"
+"Somewhat inelegantly (for the benefit of code reuse, whilst avoiding some\n"
+"harrowing partial binding for C functions) this requires that you call it\n"
+"twice, once with a `conf` list only, and again with `topic_conf` only.\n"
+"\n"
+"Repeated calls work incrementally; you can wipe configuration completely\n"
+"by calling Consumer_stop()\n";
+static PyObject *
+Consumer_configure(Consumer *self, PyObject *args, PyObject *kwds) {
+    char *keywords[] = {"conf", "topic_conf", NULL};
+    PyObject *conf = NULL;
+    PyObject *topic_conf = NULL;
+    if (! PyArg_ParseTupleAndKeywords(args,
```

Comment:
Conditionals without curly braces around their blocks can be really dangerous. I'd suggest using `{}` around the `return NULL`.